# 1. laboratorijas darbs - Teilora rinda + cietais rieksts


Dotā funkcija: y=(1+x)*exp(x)

No bibliotēkas _math_ importējam mums nepieciešamo exp funkciju:

In [3]:
from math import exp

Argumentam x  ir piešķirta vērtība 3.14 un y definējam kā __(1+x)*exp(x)__, izmantojot importēto exp funkciju un iegūstam funkcijas vērtību y(x):  

In [47]:
print("y(x)=(1+x)*e(x) funkcijas aprēķināšana:")
x = float(input("Lietotāj, lūdzu, ievadi argumentu (x): "))
y = exp(x)*(1+x)
print("y(%0.2f)=(1+%0.2f)*e(%0.2f)=%0.2f"%(x,x,x,y),"   /izmantojot bibliotēkas exp f-ju")
print()

y(x)=(1+x)*e(x) funkcijas aprēķināšana:
Lietotāj, lūdzu, ievadi argumentu (x): 3.14
y(3.14)=(1+3.14)*e(3.14)=95.65    /izmantojot bibliotēkas exp f-ju



Definējam savu funkciju un piešķiram vērtības argumentiem _k (summas locekļu skaits)_, _a (attiecīgā soļa loceklis)_, _S (summēšanā iegūtā vērtība)_ un izprintējam 0to funkcijas summēšanas locekli (yy piešķiram __mana_exp_f_ja(x)__ vērtību): 

In [37]:
def mana_exp_f_ja(x):
   k = 0
   a = ((1+x)*x**k)/(1)
   S = a
   print("a0 = %0.2f S0 = %0.2f"%(a,S))
yy = mana_exp_f_ja(x)


a0 = 4.14 S0 = 4.14


Izveidojam ciklu, kurā tiek izmantoti 1000 summas locekļi un attiecīgās f-jas aprēķinātais rekurences reizinātājs _R = x/k_, lai aprēķinātu f-jas vērtību. Izmantojot _for_, no visa cikla izprintējam tikai 999to un 1000to summēšanas vērtību (yy piešķiram __mana_exp_f_ja(x)__ vērtību): 

In [38]:
def mana_exp_f_ja(x):
   k = 0
   a = ((1+x)*x**k)/(1)
   S = a
   print("a0 = %0.2f S0 = %0.2f"%(a,S))

   while k < 1000:
      k = k + 1
      R = x / k
      a = a * R
      S = S + a
   for k in [999, 1000]:
      print("a%d = %0.2f S%d = %0.2f"%(k,a,k,S))
   return S

yy = mana_exp_f_ja(x)

a0 = 4.14 S0 = 4.14
a999 = 0.00 S999 = 95.65
a1000 = 0.00 S1000 = 95.65


Pēc summēšanas cikla izpildes 1000tais summēšanas loceklis ir 95.65, kas attiecīgi ir arī __mana_exp_f_ja__ iegūtā vērtība, to arī izprintējam: 

In [34]:
print("y(%0.2f)=(1+%0.2f)*e(%0.2f) caur summu: %0.2f"%(x,x,x,yy))
print()

y(3.14)=(1+3.14)*e(3.14) caur summu: 95.65



Pēc dotajiem norādījumiem izvadam arī summas izteiksmi un rekurences funkcionālā reizinātāja izteiksmi, izmantojot __ASCII__ simbolus: 

In [39]:
print('            1000')
print('         _________')
print('         \ ')
print('          \                       k')
print('           \         ( 1 + x ) * x')
print('(1+x)*e(x) = >     _____________________ ...,kur -\u221E < x < +\u221E')  
print('           /')
print('          /                 k!')
print('         /________')
print()
print('             k=0')
print()

print('                            x')
print('rekurences reizinātājs:  ________')
print() 
print('                            k')

            1000
         _________
         \ 
          \                       k
           \         ( 1 + x ) * x
(1+x)*e(x) = >     _____________________ ...,kur -∞ < x < +∞
           /
          /                 k!
         /________

             k=0

                            x
rekurences reizinātājs:  ________

                            k


## CIETAIS RIEKSTS

Cietajā riekstā jānoskaidro summas lockeļu skaits, kas jāizmanto, lai kļūda starp teorētisko un praktisko, ar summēšanu iegūto funkcijas tuvinājumu, būtu mazāka par 0.01. Šo locekļu skaitu ciklā jau iepriekš un arī tagad apzīmējam ar k un skaitīšanu veicam ar k=k+1. Lai to izdarītu, iepriekš lietotajā ciklā nomainam nosacījumu. Attiecīgi tagad cikls balstās uz nosacījumu __y-S >= 0.01__, tātad cikls turpinās tik ilgi, kamēr patiesās vērtības (y) un iegūtās vērtības (S) starpība ir lielāka vai vienāda ar 0.01. Kad šī starpība ir mazāka, cikls ir pabeigts. Izprintējam šo locekļu skaitu:

In [45]:
def mana_exp_f_ja(x):
   k = 0
   a = ((1+x)*x**k)/(1)
   S = a
   print("a0 = %0.2f S0 = %0.2f"%(a,S))

   while y-S >= 0.01:
      k = k + 1
      R = x / k
      a = a * R
      S = S + a
      print("a%d = %0.2f S%d = %0.2f"%(k,a,k,S))
   print('Summas locekļu skaits, kas jāizmanto, lai kļūda starp teorētisko un praktisko, ar summēšanu iegūto funkcijas tuvinājumu, būtu mazāka par 0.01 ir:',k)
   return S
yy = mana_exp_f_ja(x)

a0 = 4.14 S0 = 4.14
a1 = 13.00 S1 = 17.14
a2 = 20.41 S2 = 37.55
a3 = 21.36 S3 = 58.91
a4 = 16.77 S4 = 75.68
a5 = 10.53 S5 = 86.21
a6 = 5.51 S6 = 91.72
a7 = 2.47 S7 = 94.19
a8 = 0.97 S8 = 95.16
a9 = 0.34 S9 = 95.50
a10 = 0.11 S10 = 95.61
a11 = 0.03 S11 = 95.64
a12 = 0.01 S12 = 95.65
Summas locekļu skaits, kas jāizmanto, lai kļūda starp teorētisko un praktisko, ar summēšanu iegūto funkcijas tuvinājumu, būtu mazāka par 0.01 ir: 12
